jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ConditionalHandler.java

```diff
@@ -0,0 +1,738 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static java.lang.invoke.MethodType.methodType;
+
+/**
+ * A {@link Handler.Wrapper} that may conditionally apply an action to a request.
+ * The conditions are implemented by an {@link IncludeExclude} of:
+ * <ul>
+ *     <li>A method name, which can be efficiently matched</li>
+ *     <li>A {@link PathSpec} or sting representation, which can be efficient matched.</li>
+ *     <li>An arbitrary {@link Predicate} taking the {@link Request}, which is matched in a linear test of all predicates.</li>
+ * </ul>
+ * <p>
+ * If the conditions are met, the {@link #doHandle(Request, Response, Callback)} method will be invoked.
+ * However, as the default implementation to to call the {@link #getHandler() next Handler}, an optimization is applied to
+ * directly call the next {@code Handler} if {@code doHandler} has not been extended.
+ * </p>
+ * <p>
+ * If the conditions are not met, then the behaviour will be determined by the {@link NotHandled} passed to the
+ * constructor.
+ * </p>
+ *
+ */
+public class ConditionalHandler extends Handler.Wrapper
+{
+    private static final Logger LOG = LoggerFactory.getLogger(ConditionalHandler.class);
+
+    /**
+     * Enumeration of the actions that can be taken if the request is to not be handled due to failing to
+     * meet the conditions.
+     */
+    public enum NotHandled
+    {
+        /**
+         * If the conditions are not met, then the {@link #doNotHandle(Request, Response, Callback)} method is invoked,
+         * which by default will return {@code false}.  This action is used when the entire {@code Handler} branch is
+         * to be bypassed.
+         */
+        DO_NOT_HANDLE,
+
+        /**
+         * If the conditions are not met, then bypass the {@link #doHandle(Request, Response, Callback)} method by
+         * invoking the {@link Handler#handle(Request, Response, Callback)} method of the {@link #getHandler() next Handler}.
+         * This action is used if this class has been extended to provided specific optional behavior in the {@code doHandle}
+         * method.
+         */
+        SKIP_THIS,
+
+        /**
+         * If the conditions are not met, then bypass the {@link Handler#handle(Request, Response, Callback)} method
+         * of the {@link #getHandler() next Handler} invoking the {@link Handler#handle(Request, Response, Callback)}
+         * method of the {@link Handler} {@link Singleton#getHandler() after} the {@link #getHandler() nextHandler}.
+         * This action is used to make the specific behavior of the {@link #getHandler() following}
+         * {@link Handler.Wrapper Wrapper} class optional.
+         */
+        SKIP_NEXT,
+    }
+
+    private final NotHandled _notHandled;
+    private final IncludeExclude<String> _methods = new IncludeExclude<>();
+    private final IncludeExclude<String> _paths = new IncludeExclude<>(PathSpecSet.class);
```

Comment:
Since we're here, nothing about headers?
I.e. a match on header name/value like we do for method and path?
But can be added later.

Reply:
The problem with headers is that very often you want to extract something from the value.   You can do this by adding a general purpose predicate that puts some result in a request attribute, but hard to do generally.   Might be a good example to do eventually.